internal/impl/gcp/enterprise/input_spanner_partition_batcher.go

```diff
@@ -0,0 +1,221 @@
+// Copyright 2025 Redpanda Data, Inc.
+//
+// Licensed as a Redpanda Enterprise file under the Redpanda Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+// https://github.com/redpanda-data/connect/blob/main/licenses/rcl.md
+
+package enterprise
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"iter"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/redpanda-data/benthos/v4/public/service"
+	"github.com/redpanda-data/connect/v4/internal/ack"
+	"github.com/redpanda-data/connect/v4/internal/impl/gcp/enterprise/changestreams"
+)
+
+// spannerPartitionBatchIter goes over changestreams.DataChangeRecord.Mods,
+// for every mod it creates a message and adds it to the batch, if the batch is
+// full, it yields the batch and creates a new one.
+//
+// Iff batch is returned with nonzero time, when acked the partition watermark
+// should be updated to this time.
+type spannerPartitionBatchIter struct {
+	*spannerPartitionBatcher
+	dcr *changestreams.DataChangeRecord
+	err error
+}
+
+func (s *spannerPartitionBatchIter) Iter(ctx context.Context) iter.Seq2[service.MessageBatch, time.Time] {
+	return func(yield func(service.MessageBatch, time.Time) bool) {
+		if s.err != nil {
+			return
+		}
+
+		lastFlushed := false
+		defer func() {
+			if lastFlushed {
+				s.last = nil
+			} else {
+				s.last = s.dcr
+			}
+		}()
+
+		first := true
+		for i, m := range s.dcr.Mods {
+			b, err := json.Marshal(m)
+			if err != nil {
+				s.err = err
+				return
+			}
+
+			msg := service.NewMessage(b)
+			msg.MetaSet("table_name", s.dcr.TableName)
+			msg.MetaSet("mod_type", s.dcr.ModType)
+			msg.MetaSet("commit_timestamp", s.dcr.CommitTimestamp.Format(time.RFC3339Nano))
+			msg.MetaSet("record_sequence", s.dcr.RecordSequence)
+			msg.MetaSet("server_transaction_id", s.dcr.ServerTransactionID)
+			msg.MetaSet("is_last_record_in_transaction_in_partition", strconv.FormatBool(s.dcr.IsLastRecordInTransactionInPartition))
+			msg.MetaSet("value_capture_type", s.dcr.ValueCaptureType)
+			msg.MetaSet("number_of_records_in_transaction", strconv.FormatInt(s.dcr.NumberOfRecordsInTransaction, 10))
+			msg.MetaSet("number_of_partitions_in_transaction", strconv.FormatInt(s.dcr.NumberOfPartitionsInTransaction, 10))
+			msg.MetaSet("transaction_tag", s.dcr.TransactionTag)
+			msg.MetaSet("is_system_transaction", strconv.FormatBool(s.dcr.IsSystemTransaction))
```

Comment:
nit: use `MetaSetMut` and don't convert to string, same with other booleans and integers